virocon/contours.py

```diff
@@ -806,3 +812,266 @@ def _compute(self):
         ) / denominator
 
         self.coordinates = np.array([x_cont, y_cont]).T
+
+
+class AndContour(Contour):
+    """
+    A contour that connects points of constant AND exceedance. Such
+    contours are described, for example, in Mazas (2019) [1]_.
+    This implementation uses Monte Carlo simulation and only works for 2D models.
+
+    Parameters
+    ----------
+    model : MultivariateModel
+        The model to be used to calculate the contour.
+    alpha : float
+        The exceedance probability. The probability that an observation
+        falls outside the environmental contour.
+    n : int, optional
+        Number of data points that shall be Monte Carlo simulated. Defaults
+        to None, which calculates n based on alpha: n = int(100 / alpha).
+    deg_step : float, optional
+        Directional step in degrees. Defaults to 5.
+    sample : 2-dimensional ndarray, optional
+        Monte Carlo simulated environmental states. Array is of shape (n, d)
+        with d being the number of variables and n being the number of
+        observations.
+    allowed_error : float, optional
+        Required precision for the alpha value. For example 0.1 means that
+        the algorithm searches along the path until the probability of exceedance
+        at the current point p_e satisfies |p_e - alpha| / alpha < 0.1.
+        Defaults to 0.01.
+
+    Attributes
+    ----------
+    coordinates : ndarray
+        Coordinates of the calculated contour.
+        Shape: (number of points, number of dimensions).
+
+    References
+    ----------
+    .. [1] Mazas, F. (2019). Extreme events: a framework for assessing natural
+           hazards. Natural Hazards. https://doi.org/10.1007/s11069-019-03581-9
+
+    """
+
+    def __init__(
+        self, model, alpha, n=None, deg_step=3, sample=None, allowed_error=0.01
+    ):
+        self.model = model
+        self.alpha = alpha
+        if n is None:
+            n = int(100 / alpha)
+        self.n = n
+        self.deg_step = deg_step
+        self.sample = sample
+        self.allowed_error = allowed_error
+        super().__init__()
+
+    def _compute(self):
+        model = self.model
+        alpha = self.alpha
+        n = self.n
+        deg_step = self.deg_step
+        sample = self.sample
+        allowed_error = self.allowed_error
+
+        if self.model.n_dim != 2:
+            raise NotImplementedError(
+                "AndContour is currently only "
+                "implemented for two dimensions."
+            )
+
+        if sample is None:
+            sample = self.model.draw_sample(n)
+            self.sample = sample
+        x, y = sample.T
+
+        max_iterations = 100
```

Comment:
I added an explanation, yes, it's set empirically. I have not ran into problems with it yet such that I would not like to have it as a funciton parameter at this moment.

Reply:
If you did not run into problems, why introduce the iteration limit in the first place?

Reply:
I did not run into problems where an iteration limit of 100 is unsufficient. Such that e.g. with limit = 1000 the required precision would be reached, but not with limit = 100.

The limit is necessary, because if we have too little datapoints, we cannot reach the required precision and we need to stop the iterations that try to get a higher and higher precision at some point.

Reply:
Ah okay, I see. If we do not want to add another parameter, would it be possbile to adjust `max_iterations` depending on the accuracy the user requests? Or should we just leave it be until some user runs into issues?

Reply:
I think it's currently not worth the time thinking about how such an automatic selection could be best done. Unless you already have an idea? Then feel free to implement it :-)

Reply:
Okay, then we leave it as is.